File: built-in plugins dsp/mb-stereo-crossfeed.cpp
/**
 * MB Crossfeed
 * Category : effect
 * Type     : stereo
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Headphone crossfeed for natural speaker-like imaging
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_STEREO_CROSSFEED_H
#define MB_STEREO_CROSSFEED_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbStereoCrossfeed : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-stereo-crossfeed";
    static constexpr const char* PLUGIN_NAME    = "MB Crossfeed";
    static constexpr const char* PLUGIN_TYPE    = "stereo";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float amount = 0.3f;  // range [0, 1]
    float frequency = 700f;  // range [300, 2000]
    float delay = 0.3f;  // range [0, 1]
    };

    MbStereoCrossfeed() = default;
    ~MbStereoCrossfeed() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.amount = std::clamp(params.amount, 0f, 1f);
        params.frequency = std::clamp(params.frequency, 300f, 2000f);
        params.delay = std::clamp(params.delay, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Crossfeed
        return input;
    }
};

#endif // MB_STEREO_CROSSFEED_H

File: built-in plugins dsp/mb-stereo-crossfeed.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbStereoCrossfeedPlugin: PluginDefinition = {
  id: "mb-stereo-crossfeed",
  slug: "mb-stereo-crossfeed",
  name: "MB Crossfeed",
  category: "effect",
  type: "stereo" as any,
  version: "1.0.0",
  description: "Headphone crossfeed for natural speaker-like imaging",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "amount",
      name: "Amount",
      type: "float",
      defaultValue: 0.3,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "frequency",
      name: "Frequency",
      type: "float",
      defaultValue: 700,
      minValue: 300,
      maxValue: 2000,
      automatable: true,
    },
    {
      id: "delay",
      name: "Delay",
      type: "float",
      defaultValue: 0.3,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { amount: 0.3, frequency: 700, delay: 0.3 },
};

export default MbStereoCrossfeedPlugin;
